### PR TITLE
Add functions returning Some in Rigid.Freer

### PR DESCRIPTION
**Pull request by phadej**
EDIT: `Selective.Free` could have same functions added, I'd add them there too, if this looks good otherwise.

I played with `selective` today, especially with various free formulations. This (= that you might want to diffentiate between `<*>` and `apS`) is IMHO insightful as demonstrated in subtly hidden example:

`Some` is nice when working with GADTs, allows to `show` things:

Screenshot from 2019-10-24 22-35-43

Screenshot from 2019-10-24 22-36-05

### CODE REVIEW

File: src/Control/Selective/Rigid/Freer.hs [hunk 1]
```diff
@@ -91,11 +181,15 @@ getPure = runSelect (const Nothing)
 getEffects :: Functor f => Select f a -> [f ()]
 getEffects = foldSelect (pure . void)
 
+-- | Like 'getEffects' but not requiring @'Functor' f@.
+getEffectsSome :: Select f a -> [Some f]
```
Comment (phadej) [inline src/Control/Selective/Rigid/Freer.hs]: It's very tempting to name this `getSomeEffects`, but it would be quite misleading :)

File: src/Control/Selective/Rigid/Freer.hs [hunk 2]
```diff
@@ -91,11 +181,15 @@ getPure = runSelect (const Nothing)
...
+getEffectsSome = foldSelect (pure . mkSome)
+
 -- | Extract the necessary effect from a free selective computation. Note: there
 -- can be at most one effect that is statically guaranteed to be necessary.
 getNecessaryEffect :: Functor f => Select f a -> Maybe (f ())
-getNecessaryEffect = leftToMaybe . runExcept . runSelect (throwE . void)
+getNecessaryEffect = getFirst . foldSelect (First . Just . void)
```
Comment (snowleopard) [inline src/Control/Selective/Rigid/Freer.hs]: Nice!

File: src/Control/Selective/Rigid/Freer.hs (hunk 1 above)
Comment (snowleopard) [reply src/Control/Selective/Rigid/Freer.hs]: Indeed :)

File: src/Control/Selective/Rigid/Freer.hs [hunk 3]
```diff
@@ -23,14 +23,104 @@ module Control.Selective.Rigid.Freer (
...
 
-import Control.Monad.Trans.Except
 import Control.Selective
 import Data.Bifunctor
 import Data.Function
 import Data.Functor
+import Data.Some (Some, mkSome)
+import Data.Monoid (First (..))
+
+-- $setup
```
Comment (snowleopard) [inline src/Control/Selective/Rigid/Freer.hs]: What does `$setup` do? Is this something related to `doctest`?

**Comment by snowleopard** [comment]
@phadej Thank you, I like the API based on `Some`. What a nice little GADT!

The example is great too.

I'm a bit torn, however:
* On the one hand, I'd just replace all `f ()` functions with their `Some f` equivalents.
* On the other hand, I prefer to keep my libraries free of dependencies that are not bundled with GHC, simply because any extra dependencies tend to create non-fun maintenance work for me. (I could learn how to become better at maintaining my projects, but I'd rather learn something fun instead!)

Given that we only get to use `Some` in the fairly obscure corner of the library (`getNecessaryEffect` and `getEffects`), I'm currently leaning towards not introducing the extra dependency. I'm happy to keep `Some` in the examples: to make it work, you could just move the one-line definitions from the code to the examples. We could also advertise `Some` and alternative definitions in the docs for `getNecessaryEffect` and `getEffects`. Does this sound OK to you?

**Comment by phadej** [comment]
moving code around is not fun, compared to getting it from the dependencies. YMMV :(

**Comment by snowleopard** [comment]
@phadej If you don't mind, I'd like to merge the PR, doing the moving & copying part myself. It'd be sad to lose this example!

**Comment by phadej** [comment]
I’m completely fine with that. (You might want to add a section showing that Control.Selective.Free behaves non-rigidly, but OTOH you have to “clever” about that).

**Comment by phadej** [comment]
I don't like seeing open pull request I cannot do anything. Please not reopen. You can clone the branch, if you want it.

**Comment by snowleopard** [comment]
Sure, hope to find time for this soon.
